power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
-    std::vector<EdgeWeight> edge_weight(boost::num_vertices(graph), infty);
-    std::vector<Idx2D> edge_pos(boost::num_vertices(graph));
-
-    for (auto v : boost::make_iterator_range(boost::vertices(graph))) {
+inline auto get_edge_weights(TransformerGraph const& graph) -> TrafoGraphEdgeProperties {
+    std::vector<EdgeWeight> vertex_distances(boost::num_vertices(graph), infty);
+    BGL_FORALL_VERTICES(v, graph, TransformerGraph) {
         if (graph[v].is_source) {
-            process_edges_dijkstra(v, edge_weight, edge_pos, graph);
+            process_edges_dijkstra(v, vertex_distances, graph);
         }
     }
 
-    WeightedTrafoList result;
-    for (size_t i = 0; i < edge_weight.size(); ++i) {
-        result.emplace_back(edge_pos[i], edge_weight[i]);
+    TrafoGraphEdgeProperties result;
+    BGL_FORALL_EDGES(e, graph, TransformerGraph) {
+        if (graph[e].regulated_idx == unregulated_idx) {
+            continue;
+        }
+        // result.emplace_back(graph[e].regulated_idx, vertex_distances[boost::source(e, graph)]);
+        result.push_back({graph[e].regulated_idx, vertex_distances[boost::source(e, graph)]});
     }
 
     return result;
 }
 
-// Step 4: Loop all transformers with automatic tap changers, including the transformers which are not
-//         fully connected
-//   a.Rank of the transformer <-
-//       i. Infinity(INT_MAX), if tap side of the transformer is disconnected.
-//          The transformer regulation should be ignored
-//       ii.Rank of the vertex at the tap side of the transformer, if tap side of the transformer is connected
-inline auto transformer_disconnected(Idx2D const& /*pos*/) -> bool {
-    // <TODO: jguo> waiting for the functionalities in step 1 to be implemented
-    return false;
-}
-
-inline auto rank_transformers(WeightedTrafoList const& w_trafo_list) -> RankedTransformerGroups {
+inline auto rank_transformers(TrafoGraphEdgeProperties const& w_trafo_list) -> RankedTransformerGroups {
     auto sorted_trafos = w_trafo_list;
 
-    for (auto& trafo : sorted_trafos) {
-        if (transformer_disconnected(trafo.first)) {
-            trafo.second = infty;
-        }
-    }
-
     std::sort(sorted_trafos.begin(), sorted_trafos.end(),
-              [](const WeightedTrafo& a, const WeightedTrafo& b) { return a.second < b.second; });
-
-    RankedTransformerGroups groups;
-    Idx last_weight = -1;
-    for (const auto& trafo : sorted_trafos) {
-        if (groups.empty() || last_weight != trafo.second) {
-            groups.push_back(std::vector<Idx2D>{trafo.first});
-            last_weight = trafo.second;
-        } else {
-            groups.back().push_back(trafo.first);
-        }
-    }
+              [](const TrafoGraphEdge& a, const TrafoGraphEdge& b) { return a.weight < b.weight; });
+
+    RankedTransformerGroups groups(sorted_trafos.size());
+    std::ranges::transform(sorted_trafos.begin(), sorted_trafos.end(), groups.begin(),
```

Comment:
if you use `begin` and `end` but don't make use of the `|` functionality of `std::ranges` then you might as well use `std::ranges` directly

```suggestion
    std::transform(sorted_trafos.begin(), sorted_trafos.end(), groups.begin(),
```

Reply:
That change is thanks to sonar cloud nagging :D

Reply:
OHHHHH i understand why. this should be
```suggestion
    std::ranges::transform(sorted_trafos, groups.begin(),
```

cause `RankedTransformerGroups` is an `std::vector` and therefore supports direct range access

Reply:
The change is really not adding much, but since you typed so many letters already on this, I'll remove the 28 letters